crates/ruff/src/rules/refurb/rules/repeated_append.rs

```diff
@@ -0,0 +1,442 @@
+use rustc_hash::FxHashMap;
+
+use ast::{traversal, ParameterWithDefault, Parameters};
+use ruff_diagnostics::{AutofixKind, Diagnostic, Edit, Fix, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::helpers::map_subscript;
+use ruff_python_ast::{self as ast, Expr, Stmt};
+use ruff_python_codegen::Generator;
+use ruff_python_semantic::analyze::type_inference::{PythonType, ResolvedPythonType};
+use ruff_python_semantic::{Binding, BindingId, BindingKind, DefinitionId, SemanticModel};
+use ruff_text_size::{Ranged, TextRange};
+
+use crate::autofix::snippet::SourceCodeSnippet;
+use crate::checkers::ast::Checker;
+use crate::registry::AsRule;
+
+/// ## What it does
+/// Checks for consecutive calls to `append`.
+///
+/// ## Why is this bad?
+/// Consecutive calls to `append` can be less efficient than batching them into
+/// a single `extend`. Each `append` resizes the list individually, whereas an
+/// `extend` can resize the list once for all elements.
+///
+/// ## Example
+/// ```python
+/// nums = [1, 2, 3]
+///
+/// nums.append(4)
+/// nums.append(5)
+/// nums.append(6)
+/// ```
+///
+/// Use instead:
+/// ```python
+/// nums = [1, 2, 3]
+///
+/// nums.extend((4, 5, 6))
+/// ```
+///
+/// ## References
+/// - [Python documentation: More on Lists](https://docs.python.org/3/tutorial/datastructures.html#more-on-lists)
+#[violation]
+pub struct RepeatedAppend {
+    name: String,
+    replacement: SourceCodeSnippet,
+}
+
+impl RepeatedAppend {
+    fn suggestion(&self) -> String {
+        let name = &self.name;
+        self.replacement
+            .full_display()
+            .map_or(format!("{name}.extend(...)"), ToString::to_string)
+    }
+}
+
+impl Violation for RepeatedAppend {
+    const AUTOFIX: AutofixKind = AutofixKind::Sometimes;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let name = &self.name;
+        let suggestion = self.suggestion();
+        format!("Use `{suggestion}` instead of repeatedly calling `{name}.append()`")
+    }
+
+    fn autofix_title(&self) -> Option<String> {
+        let suggestion = self.suggestion();
+        Some(format!("Replace with `{suggestion}`"))
+    }
+}
+
+/// FURB113
+pub(crate) fn repeated_append(checker: &mut Checker, stmt: &Stmt) {
+    let Some(appends) = match_consecutive_appends(checker.semantic(), stmt) else {
+        return;
+    };
+
+    // No need to proceed if we have less than 1 `append` to work with.
+    if appends.len() <= 1 {
+        return;
+    }
+
+    // group borrows from checker, so we can't directly push into checker.diagnostics
+    let diagnostics: Vec<Diagnostic> = group_appends(appends)
+        .iter()
+        .filter_map(|group| {
```

Comment:
Writing this as a map felt a bit more idiomatic to me.

Reply:
Looks better, agree!